### PR TITLE
[macOS] Adopt RunningBoard process assertions

### PR DESCRIPTION
**Pull request by salinas-miguel**
#### 1f756e6601a3cfd67cb7a240fa4ee53e03dbcb98
<pre>
[macOS] Adopt RunningBoard process assertions
<a href="https://bugs.webkit.org/show_bug.cgi?id=244998">https://bugs.webkit.org/show_bug.cgi?id=244998</a>
rdar://99753987

Reviewed by Chris Dumez.

Adopt RunningBoard process assertions on macOS and stop leaking an
os_transaction in order to keep our XPC services alive.
This paves the way for suspendable processes on macOS since we now
receive SIGKILL on application exit instead of SIGTERM.

* Source/WTF/wtf/PlatformUse.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Scripts/update-info-plist-for-runningboard.sh: Added.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
(WebKit::XPCServiceInitializer):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::XPCServiceExit):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm: Renamed from Source/WebKit/UIProcess/ios/ProcessAssertionIOS.mm.
(assertionsWorkQueue):
(processHasActiveRunTimeLimitation):
(+[WKProcessAssertionBackgroundTaskManager shared]):
(-[WKProcessAssertionBackgroundTaskManager init]):
(-[WKProcessAssertionBackgroundTaskManager dealloc]):
(-[WKProcessAssertionBackgroundTaskManager addAssertionNeedingBackgroundTask:]):
(-[WKProcessAssertionBackgroundTaskManager removeAssertionNeedingBackgroundTask:]):
(-[WKProcessAssertionBackgroundTaskManager _notifyAssertionsOfImminentSuspension]):
(-[WKProcessAssertionBackgroundTaskManager _scheduleReleaseTask]):
(-[WKProcessAssertionBackgroundTaskManager _cancelPendingReleaseTask]):
(-[WKProcessAssertionBackgroundTaskManager _hasBackgroundTask]):
(-[WKProcessAssertionBackgroundTaskManager _updateBackgroundTask]):
(-[WKProcessAssertionBackgroundTaskManager assertionWillInvalidate:]):
(-[WKProcessAssertionBackgroundTaskManager assertion:didInvalidateWithError:]):
(-[WKProcessAssertionBackgroundTaskManager _handleBackgroundTaskExpiration]):
(-[WKProcessAssertionBackgroundTaskManager _handleBackgroundTaskExpirationOnMainThread]):
(-[WKProcessAssertionBackgroundTaskManager _releaseBackgroundTask]):
(-[WKProcessAssertionBackgroundTaskManager setProcessStateMonitorEnabled:]):
(-[WKRBSAssertionDelegate dealloc]):
(-[WKRBSAssertionDelegate assertionWillInvalidate:]):
(-[WKRBSAssertionDelegate assertion:didInvalidateWithError:]):
(WebKit::runningBoardNameForAssertionType):
(WebKit::runningBoardDomainForAssertionType):
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::remainingRunTimeInSeconds):
(WebKit::ProcessAssertion::acquireAsync):
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
(WebKit::ProcessAssertion::processAssertionWillBeInvalidated):
(WebKit::ProcessAssertion::processAssertionWasInvalidated):
(WebKit::ProcessAssertion::isValid const):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
(WebKit::ProcessAndUIAssertion::~ProcessAndUIAssertion):
(WebKit::ProcessAndUIAssertion::updateRunInBackgroundCount):
(WebKit::ProcessAndUIAssertion::setProcessStateMonitorEnabled):
(WebKit::ProcessAndUIAssertion::uiAssertionWillExpireImminently):
(WebKit::ProcessAndUIAssertion::processAssertionWasInvalidated):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::shouldLeakBoost):
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/254759@main">https://commits.webkit.org/254759@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3a428457727e2c6ed1b54e47c4f092832e76e73f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34566 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99358 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156863 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33058 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28442 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82356 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93653 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95667 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26272 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76833 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26190 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/93599 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81133 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80934 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69190 "Found 4 new API test failures: /WebKit2Gtk/TestLoaderClient:/webkit/WebKitWebPage/redirect-to-data-uri, /TestWTF:WTF_WordLock.ManyContendedLongSections, /WebKit2Gtk/TestWebKitVersion:/webkit/WebKitVersion/version, /WebKit2Gtk/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/children-changed (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81647 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15017 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76535 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15958 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26487 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3343 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38921 "Found 1 new test failure: fast/events/blur-remove-parent-crash.html (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79121 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35042 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17337 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.no-llint (failure)") |

### CODE REVIEW

File: Source/WebKit/UIProcess/ProcessAssertion.h [hunk 1]
```diff
@@ -87,20 +87,16 @@ class ProcessAssertion : public ThreadSafeRefCounted<ProcessAssertion>, public C
     void acquireAsync(CompletionHandler<void()>&&);
     void acquireSync();
 
-#if PLATFORM(IOS_FAMILY)
```
Comment (salinas-miguel) [inline Source/WebKit/UIProcess/ProcessAssertion.h:90]: I made this mistake in a few places. We still need the platform check but it should be COCOA now. I'll update it now.

File: Source/WebKit/UIProcess/mac/WebPageProxyMac.mm [hunk 2]
```diff
@@ -680,6 +680,27 @@ static inline bool expectsLegacyImplicitRubberBandControl()
 
 #endif // ENABLE(UI_PROCESS_PDF_HUD)
 
+void WebPageProxy::processWillBecomeSuspended()
+{
+    if (!hasRunningProcess())
+        return;
+
+//    m_hasNetworkRequestsOnSuspended = m_pageLoadState.networkRequestsInProgress();
```
Comment (salinas-miguel) [inline Source/WebKit/UIProcess/mac/WebPageProxyMac.mm]: Oops. I'll fix this too.

**Comment by salinas-miguel** [comment]
Looking into the test crashes

File: Source/WebKit/Scripts/update-info-plist-for-runningboard.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+if [[ "${WK_PLATFORM_NAME}" == macosx ]] && (( "${TARGET_MAC_OS_X_VERSION_MAJOR}" >= 130000 ))
+then
+    /usr/libexec/PlistBuddy -c "Add :LSDoNotSetTaskPolicyAutomatically bool YES" "${SCRIPT_INPUT_FILE_0}"
+    /usr/libexec/PlistBuddy -c "Add :_AdditionalProperties: dict" "${SCRIPT_INPUT_FILE_0}"
+    /usr/libexec/PlistBuddy -c "Add :_AdditionalProperties:RunningBoard: dict" "${SCRIPT_INPUT_FILE_0}"
+    /usr/libexec/PlistBuddy -c "Add :_AdditionalProperties:RunningBoard:Managed bool YES" "${SCRIPT_INPUT_FILE_0}"
+    /usr/libexec/PlistBuddy -c "Add :_AdditionalProperties:RunningBoard:Reported bool YES" "${SCRIPT_INPUT_FILE_0}" 
+fi
```
Comment (salinas-miguel) [inline Source/WebKit/Scripts/update-info-plist-for-runningboard.sh]: This script isn't doing what I thought it did. Looking into it.

Comment (salinas-miguel) [reply Source/WebKit/Scripts/update-info-plist-for-runningboard.sh]: Fixed

File: Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h [hunk 4]
```diff
@@ -30,7 +30,9 @@
 #import <JavaScriptCore/ExecutableAllocator.h>
 #import <wtf/OSObjectPtr.h>
 #import <wtf/cocoa/RuntimeApplicationChecksCocoa.h>
+#if !USE(RUNNINGBOARD)
```
Comment (cdumez) [inline Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:34]: nit: We usually have a blank line before conditional imports.

File: Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp [hunk 5]
```diff
@@ -288,6 +288,10 @@ void AuxiliaryProcessProxy::didFinishLaunching(ProcessLauncher*, IPC::Connection
     if (!connectionIdentifier)
         return;
 
+#if PLATFORM(MAC) && USE(RUNNINGBOARD)
+    m_lifetimeAssertion = ProcessAssertion::create(xpc_connection_get_pid(connectionIdentifier.xpcConnection.get()), "Lifetime assertion"_s, ProcessAssertionType::Foreground);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:292]: Isn't a background assertion enough here? Presumably, if the tab is foreground, the process throttler will take its own foreground assertion and this is just the backup assertion to make sure we never suspend on macOS?

File: Source/WebKit/UIProcess/AuxiliaryProcessProxy.h [hunk 6]
```diff
@@ -28,6 +28,7 @@
 #include "AuxiliaryProcessCreationParameters.h"
 #include "Connection.h"
 #include "MessageReceiverMap.h"
+#include "ProcessAssertion.h"
```
Comment (cdumez) [inline Source/WebKit/UIProcess/AuxiliaryProcessProxy.h]: A forward declaration should suffice.

File: Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm [hunk 7]
```diff
@@ -455,6 +462,25 @@ - (void)assertion:(RBSAssertion *)assertion didInvalidateWithError:(NSError *)er
...
 
+ProcessAndUIAssertion::ProcessAndUIAssertion(pid_t pid, const String& reason, ProcessAssertionType assertionType)
+    : ProcessAssertion(pid, reason, assertionType)
+{
+#if PLATFORM(IOS_FAMILY)
+    updateRunInBackgroundCount();
+#endif
+}
+
+#if PLATFORM(IOS_FAMILY)
```
Comment (cdumez) [inline Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm]: nit: I would have just guarded the destructor implementation to make the code a bit more concise.

**Review by cdumez — APPROVED** [review]
LGTM with nits.

File: Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp (hunk 5 above)
Comment (salinas-miguel) [reply Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:292]: Good question. I'm not enabling the process throttler yet with this change. Instead I'm taking a foreground assertion on each process for the entirety of its process. This is nearly equivalent to what we do currently with leaking an os_transaction/boost. The reason is that I want to preserve existing behavior as much as possible with this change (note: this preserves App Nap).

Comment (cdumez) [reply Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:292]: Ok. Sounds good.

**Comment by salinas-miguel** [comment]
@cdumez I addressed the nits but strangely I'm now seeing a failure on blur-remove-parent-crash.html on mac-AS-debug-wk2 only. It's not reproducing locally. Also there are linux failures now which is surprising given that this should only be touching Cocoa code. Could these be unrelated to my changes?

**Comment by cdumez** [comment]
> @cdumez I addressed the nits but strangely I'm now seeing a failure on blur-remove-parent-crash.html on mac-AS-debug-wk2 only. It's not reproducing locally. Also there are linux failures now which is surprising given that this should only be touching Cocoa code. Could these be unrelated to my changes?

They look unrelated to your change.

**Pull request merged**